### PR TITLE
useradd.c: fix covscan RESOURCE_LEAK

### PR DESCRIPTION
**Pull request by ikerexxe**
Error: RESOURCE_LEAK (CWE-772): [#def28]
shadow-4.8.1/src/useradd.c:1905: open_fn: Returning handle opened by "open". [Note: The source code implementation of the function has been overridden by a user model.]
shadow-4.8.1/src/useradd.c:1905: var_assign: Assigning: "fd" = handle returned from "open("/var/log/faillog", 2)".
shadow-4.8.1/src/useradd.c:1906: noescape: Resource "fd" is not freed or pointed-to in "lseek".
shadow-4.8.1/src/useradd.c:1917: leaked_handle: Handle variable "fd" going out of scope leaks the handle.
 1915|   		/* continue */
 1916|   	}
 1917|-> }
 1918|
 1919|   static void lastlog_reset (uid_t uid)

Error: RESOURCE_LEAK (CWE-772): [#def29]
shadow-4.8.1/src/useradd.c:1938: open_fn: Returning handle opened by "open". [Note: The source code implementation of the function has been overridden by a user model.]
shadow-4.8.1/src/useradd.c:1938: var_assign: Assigning: "fd" = handle returned from "open("/var/log/lastlog", 2)".
shadow-4.8.1/src/useradd.c:1939: noescape: Resource "fd" is not freed or pointed-to in "lseek".
shadow-4.8.1/src/useradd.c:1950: leaked_handle: Handle variable "fd" going out of scope leaks the handle.
 1948|   		/* continue */
 1949|   	}
 1950|-> }
 1951|
 1952|   static void tallylog_reset (const char *user_name)

Error: RESOURCE_LEAK (CWE-772): [#def30]
shadow-4.8.1/src/useradd.c:2109: alloc_fn: Storage is returned from allocation function "strdup".
shadow-4.8.1/src/useradd.c:2109: var_assign: Assigning: "bhome" = storage returned from "strdup(prefix_user_home)".
shadow-4.8.1/src/useradd.c:2131: noescape: Resource "bhome" is not freed or pointed-to in "strtok".
shadow-4.8.1/src/useradd.c:2207: leaked_storage: Variable "bhome" going out of scope leaks the storage it points to.
 2205|   		}
 2206|   #endif
 2207|-> 	}
 2208|   }
 2209|

### CODE REVIEW

File: src/useradd.c [hunk 1]
```diff
@@ -1997,17 +2002,22 @@ static void lastlog_reset (uid_t uid)
...
-	    || (fsync (fd) != 0)
-	    || (close (fd) != 0)) {
+	    || (fsync (fd) != 0)) {
 		fprintf (stderr,
 		         _("%s: failed to reset the lastlog entry of UID %lu: %s\n"),
 		         Prog, (unsigned long) uid, strerror (errno));
 		SYSLOG ((LOG_WARN, "failed to reset the lastlog entry of UID %lu", (unsigned long) uid));
-		/* continue */
 	}
+	close (fd);
```
Comment (hallyn) [inline src/useradd.c:2020]: This looks good, but please still check the close() return value.

Comment (hallyn) [reply src/useradd.c:2020]: (And yes I know I merged the other one with that check :)

Comment (ikerexxe) [reply src/useradd.c]: Done

Comment (ikerexxe) [reply src/useradd.c]: By the way, I've also changed the same error on `faillog_reset()`.

File: src/useradd.c [hunk 2]
```diff
@@ -1964,16 +1964,26 @@ static void faillog_reset (uid_t uid)
...
+	if (   (lseek (fd, offset_uid, SEEK_SET) != offset_uid)
 	    || (write (fd, &fl, sizeof (fl)) != (ssize_t) sizeof (fl))
-	    || (fsync (fd) != 0)
-	    || (close (fd) != 0)) {
+	    || (fsync (fd) != 0)) {
 		fprintf (stderr,
 		         _("%s: failed to reset the faillog entry of UID %lu: %s\n"),
 		         Prog, (unsigned long) uid, strerror (errno));
 		SYSLOG ((LOG_WARN, "failed to reset the faillog entry of UID %lu", (unsigned long) uid));
-		/* continue */
```
Comment (hallyn) [inline src/useradd.c:1976]: Sorry for the nit (of sorts), but originally all these cases continued on error.  In the first case you are now return()ing, and in the others you are not returning but dropped the comment making clear that continuing was a conscious decision.

What do you think, should we keep the comment and continue in all cases?  Change the behavior?

Comment (ikerexxe) [reply src/useradd.c:1976]: IMO, we should return if the file is not opened as I don't think it makes any sense to operate on a file that hasn't been opened. In any other circumstance we should continue and I've updated the code accordingly to include the comment again.

Comment (hallyn) [reply src/useradd.c:1976]: Huh. I was thinking this was happening in a loop.  Right you are.  Thanks.

**Review by hallyn — APPROVED** [review]

**Pull request merged**